traffic_ops/traffic_ops_golang/server/servers.go

```diff
@@ -139,94 +366,213 @@ func (s *TOServer) Validate() error {
+			if addr.Gateway != nil {
+				if gateway := net.ParseIP(*addr.Gateway); gateway == nil {
+					errs = append(errs, fmt.Errorf("%s: gateway: could not parse '%s' as a network gateway", ruleName, *addr.Gateway))
+				} else if (gateway.To4() == nil && parsedIP.To4() != nil) || (gateway.To4() != nil && parsedIP.To4() == nil) {
+					errs = append(errs, errors.New(ruleName+": address family mismatch between address and gateway"))
+				}
+			}
 
-	return message, nil
+			if addr.ServiceAddress {
+				if serviceInterface != "" && serviceInterface != iface.Name {
+					errs = append(errs, fmt.Errorf("interfaces: both %s and %s interfaces contain service addresses - only one service-address-containing-interface is allowed", serviceInterface, iface.Name))
+				}
+				serviceInterface = iface.Name
+				if parsedIP.To4() != nil {
+					if serviceAddrV4Found {
+						errs = append(errs, fmt.Errorf("interfaces: address '%s' of interface '%s' is marked as a service address, but an IPv4 service address appears earlier in the list", addr.Address, iface.Name))
+					}
+					serviceAddrV4Found = true
+				} else {
+					if serviceAddrV6Found {
+						errs = append(errs, fmt.Errorf("interfaces: address '%s' of interface '%s' is marked as a service address, but an IPv6 service address appears earlier in the list", addr.Address, iface.Name))
+					}
+					serviceAddrV6Found = true
+				}
+			}
+		}
+	}
+
+	errs = append(errs, validateCommon(s.CommonServerProperties, tx)...)
+	return serviceInterface, util.JoinErrs(errs)
 }
 
-func (s *TOServer) Read() ([]interface{}, error, error, int) {
-	version := s.APIInfo().Version
-	if version == nil {
-		return nil, nil, errors.New("TOServer.Read called with nil API version"), http.StatusInternalServerError
+func Read(w http.ResponseWriter, r *http.Request) {
+	inf, userErr, sysErr, errCode := api.NewInfo(r, nil, nil)
+	tx := inf.Tx.Tx
+	if userErr != nil || sysErr != nil {
+		api.HandleErr(w, r, tx, errCode, userErr, sysErr)
+		return
 	}
+	defer inf.Close()
 
-	returnable := []interface{}{}
+	// Middleware should've already handled this, so idk why this is a pointer at all tbh
+	version := inf.Version
+	if version == nil {
+		middleware.NotImplementedHandler().ServeHTTP(w, r)
+		return
+	}
 
-	servers, userErr, sysErr, errCode := getServers(s.ReqInfo.Params, s.ReqInfo.Tx, s.ReqInfo.User)
+	servers := []tc.ServerNullable{}
+	var unfiltered uint64
+	servers, unfiltered, userErr, sysErr, errCode = getServers(inf.Params, inf.Tx, inf.User)
 
 	if userErr != nil || sysErr != nil {
-		return nil, userErr, sysErr, errCode
+		api.HandleErr(w, r, tx, errCode, userErr, sysErr)
+		return
+	}
+
+	if version.Major >= 3 {
+		api.WriteRespWithSummary(w, r, servers, unfiltered)
+		return
 	}
 
+	if version.Major <= 1 {
+		legacyServers := make([]tc.ServerNullableV11, 0, len(servers))
+		for _, server := range servers {
+			legacyServer, err := server.ToServerV2()
```

Comment:
There's no valid reason why a _valid_ V3 server  can't be converted to a V2 or V1 server. Why?

Reply:
Just wanted to make sure, otherwise someone could submit an invalid v3 server that would cause the v1/v2 APIs to break and become unavailable.

Reply:
Yeah, that should be impossible, barring direct database manipulation.

Reply:
I guess also I should mention that even through direct database manipulation you can't _break_ the V1/V2 APIs, just make them not return IPAddress values even if the server has configured addresses - though they'd be non-service addresses so that's still technically correct.